fastai/widgets/image_downloader.py

```diff
@@ -0,0 +1,206 @@
+                                        layout=Layout(width='60px'))
+        self._size_input = Dropdown(options= ['>400*300', '>640*480', '>800*600', '>1024*768', '>2MP', '>4MP', '>6MP', '>8MP', '>10MP'],
+                                    value='>400*300',
+                                    layout=Layout(width='120px'))
+        self._download_button = Button(description="Search & Download", icon="download", layout=Layout(width='200px'))
+        self._download_button.on_click(self.on_download_button_click)
+
+        self._output = Output()
+
+        # Top horizontal controls bar
+        self._controls_pane  = HBox([self._search_input,
+                                    self._count_input,
+                                    self._size_input,
+                                    self._download_button],
+                                    layout=Layout(width='auto', height='40px'))
+
+        self._heading = ""
+        self._download_complete_heading = "<h3>Download complete. Here are a few images</h3>"
+        self._preview_header = widgets.HTML(self._heading, layout=Layout(height='60px'))
+        self._img_pane = Box(layout=Layout(display='inline'))
+        return VBox([self._controls_pane, self._preview_header, self._img_pane])
+
+
+    def render(self) -> None:
+        "Render the image search widget."
+        clear_output()
+        display(self._ui)
+
+    def clear_imgs(self) -> None:
+        "Clear the widget's images preview pane."
+        self._preview_header.value = self._heading
+        self._img_pane.children = tuple()
+
+    def validate_search_input(self) -> bool:
+        "Check if input value is empty."
+        input = self._search_input
+        if input.value == str():
+            input.layout = Layout(border="solid 2px red", height='auto')
+        else:
+            self._search_input.layout = Layout()
+        return input.value != str()
+
+    def on_download_button_click(self, btn) -> None:
+        "Download button click handler: validate search term and download images."
+        term = self._search_input.value
+        limit = int(self._count_input.value)
+        size = self._size_input.value
+
+        if not self.validate_search_input(): return
+
+        self.clear_imgs()
+        downloaded_images = download_google_images(self._path, term, n_images=limit, size=size)
+        self.display_images_widgets(downloaded_images[:min(limit, 12)])
+        self._preview_header.value = self._download_complete_heading
+        self.render()
+
+    def display_images_widgets(self, fnames:list) -> None:
+        "Display a few preview images in the notebook"
+        imgs = [widgets.Image(value=open(f, 'rb').read(), width='200px') for f in fnames]
+        self._img_pane.children = tuple(imgs)
+
+
+def download_google_images(path:PathOrStr, search_term:str, size:str='>400*300', n_images:int=10, format:str='jpg',
+                            max_workers:int=8, timeout:int=4) -> FilePathList:
```

Comment:
I'd make max_workers defaut to defaults.cpus (so a None in the default arg then that value passed in a test inside the function that uses it, which seems to be _download_images).
Also we prefer to have just one line for the docstring and the full documentation on the corresponding doc notebook.

Reply:
Stripped extra docstrings in public functions (that are documented in notebooks). Left some extra docs in private functions, since those might not be documented otherwise. 

I agree, using `defaults.cpus` is a good idea and exactly the purpose of having `defaults.cpus` in the first place, thanks for pointing that out!